### ISSUE TITLE
Выход за границу массива

### ISSUE THREAD

**Issue by gill0bates** [ProhorP/1C-inversion#1]
Вываливается с out of range - говорят это типовый способ защиты от реврсинга.
Есть где то информация, что с этим делать?

**Comment by ProhorP** [comment]
Здравствуйте, информации нет. Я переводил закрытые общие модули содержащие 10000 строк - ошибок не было. С вашим случаем не сталкивался.

**Comment by gill0bates** [comment]
Могу скинуть пример bin файла (260 строк), пока сам не пойму в чем причина, возможно в операторах jmp.

**Comment by ProhorP** [comment]
скидывайте

**Comment by gill0bates** [comment]
[Module.zip](https://github.com/ProhorP/1C-inversion/files/9971626/Module.zip)

Вот пример файла. Есть пару вопросов - писать более некуда пишу сюда.

1) Почему обработка распознает jmp как условные операторы, а не как просто переходы?
2) В строке 31 инструкция {3,8} - загрузить в переменную с индексом 8 (здесь в принципе и падает) - в данной процедуре все лишь три локальных переменных, как ему удается обратится к 8-ой при выполнении?
3) Как определюются границы процедур и локальные переменные которые им принадлежат? Не влияют ли безусловные переходы на доступность локальных переменныз?
4) Что делать с командами, которые не поддерживает 1С (например 41 - переход если истина?

**Comment by ProhorP** [comment]
1. Просто переходы определяются другим оператором, в коде обработки можешь посмотреть ее номер.
2. Посмотрю по возможности, прямо сейчас нет времени, но планирую глянуть.
3. Границы процедур определяются каки-м то номером. Посмотри в коде обработки "КонецПроцедуры" - там увидишь.
4. Такие команды никак не обрабатываются обработкой - нужно самому тебе придумать поведение при 41 операторе.

**Comment by gill0bates** [comment]
Огромное спасибо за такой эпохальный труд - буду копать.

**Comment by ProhorP** [comment]
Посмотрел ваш модуль, его отдельно видимо корректировали, даже возможно через подобную обработку. Используются метки "Перейти" в соседние процедуры, что невозможно ни в одном языке программирования кроме ассемблера :) Там всего 260 строк, это значит что примерно около 100 строк кода. Я бы на вашем месте либо бросил затею, либо попытался бы проследить все прыжки (оператор 42) и переход на заданный шаг(39) и так оследить из каких процедур в какие переместили код. Оператор 39 очень сложный, он является дополнением к проверкам условий ("Если", "Пока" "Для" "Для каждого" "Исключение" и другие, смотрите в модуле к каким операторам он привязан), но он всегда является как бы границей любого условного ветвления, он не используется сам по себе. Если у вас получится вычислить, что он не относится к условному ветвлению - значит он не является частью кода 1С, а просто является указателем на следующий участок кода. Больше я вам посоветовать ничего не могу. Но обязательно прочитайте статью от изначального автора https://habr.com/ru/post/489392/. У меня на написание обратного преобразования кода ушло 3 недели по 8 часов в день.

**Comment by gill0bates** [comment]
Спасибо за развернутый ответ, я думал я что-то не понимаю просто, теперь понятно, что на самом деле именно так.
Статья конечно я эту читал самой первой, оттуда по кросссылка и нашел Ваш гитхаб.
Эти 200 байт специально скинул как пример - чтобы немного так процедура, которая мне нужна 62000 строк.

Самый прям такой вопрос который мне неясено от слова совсем - пресловутая инструкция {3,8} - обращение к другой локальной переменной из модуля где этой переменной нет.

**Comment by ProhorP** [comment]
До {3,8} кажется был переход в середину другой процедуры. у 3-й процедуры есть такое количество переменных.

**Comment by gill0bates** [comment]
Да, но мы же находимся в контексте первой процедуры (когда возвращаемся из перехода) или сам переход дает доступ к этим переменным?

**Comment by ProhorP** [comment]
вот тут я вам не отвечу, не знаю.
